Refactor sink rendering code, move closely coupled code together in one place

aljoscha opened the pull request:
The main goal was to move any code that depends on a specific sink out of the main sink rendering method and closer to the specific sink code.

To do this, I'm introducing a `SinkRender` trait that consists of methods that allow a sink to announce what it needs to the generic rendering code, along with a `render` method that will create the sink-dependent dataflow operators.

This is mostly moving code around and extracting methods.

Two bigger changes are:
 - I removed the `TAIL` envelope because it's not a real envelope in the sense of the other envelopes
 - I simplified the `TAIL` code, removed an unnecessary `consolidate()`, added another unit test for `TAIL`

The fact that the `TAIL` code could be simplified this way was obscured before by the splitting up of the `TAIL` code into a pseudo envelope and the "real" sink code.

All the changes are heavily documented in the commits.

This is by no means the desired end state but I think it's a step in the right direction. Eventually, I think it should be possible for sink implementors to implement a trait, without having to fiddle with writing actual dataflow code.

JLDLaughlin commented on src/dataflow/src/render/sinks.rs:
```diff
@@ -45,8 +45,6 @@ where
         import_ids: HashSet<GlobalId>,
         sink_id: GlobalId,
         sink: &SinkDesc,
-        _worker_index: usize,
```
🚮

JLDLaughlin commented on src/dataflow/src/render/sinks.rs:
```diff
@@ -241,6 +217,25 @@ fn render_kafka_sink<G>(
 where
     G: Scope<Timestamp = Timestamp>,
 {
+    // consisten/exactly-once Kafka sinks need the timestamp in the row
```
nit: typo in `consistent`

aljoscha replied:
👌

JLDLaughlin commented on src/dataflow/src/render/sinks.rs:
```diff
@@ -158,23 +158,7 @@ where
-                    }
-                })
-                .as_collection(),
+            None => keyed.map(|(key, value)| (key, Some(value))),
```
I totally get wanting to pull the fake Tail envelope out, but without your git commit comment it's hard to understand what `None` would imply here. It might be helpful to add an inline comment?

aljoscha replied:
I'll add a comment. 👍

JLDLaughlin commented on src/dataflow/src/sink/tail.rs:
```diff
@@ -56,18 +68,26 @@ pub fn tail<G>(
                     while cursor.val_valid(&batch) {
                         let row = cursor.val(&batch);
                         cursor.map_times(&batch, |time, diff| {
-                            assert!(*diff >= 0, "negative multiplicities sinked in tail");
```
Did removing the `consolidate` call get rid of the chance of negative multiplicities?

aljoscha replied:
Even before, there was no chance of negative multiplicities. Except maybe with CDCv2 sources? 🤔 I'll try that.

But it's probably best not to remove the assert, I'll undo that.

aljoscha replied:
Scratch that, I now remembered the real reason. I'm adding this as a comment on the commit:

Expanding a bit, the unrolled pipeline for TAIL, combining the
"envelope" and the sink code, was roughly as shown below. I'm taking
some artistic licence here, using Streams where the actual code as
Collections, to show the types more clearly, and other shortcuts:

```
let input: Stream<(Row, time, diff)> = ...
let enriched: Stream<(Row, time, 1)> = input
  .consolidate()
  .map(|(row, time, diff)| {
    let wrapping_row = Row::new();
    if (emit_progress) {
      wrapping_row.push(false);
    }
    wrapping_row.push(time);
    wrapping_row.push(diff);
    wrapping_row.push(row);
    (<some key>, (wrapping_row, time, 1))
  }
  .arrange_by_key();

sink::tail(enriched);
```

So the actual rows, along with their time/diff where put into wrapping
rows. And tail() would write out those rows.

Now, we lose the first consolidate() call, have one arrange_by_key(),
and then write out the user/vanilla rows straight from that, and create
the wrapping Row right in the sink:tail() code, without sending it along
Streams.

The assert! that we removed in sink::tail() was asserting on the diff of
the wrapping row. Now we straight get the vanilla rows, which can (and
should) have negative multiplicities.

JLDLaughlin replied:
Awesome, that makes sense!

JLDLaughlin commented on src/dataflow/src/sink/tail.rs:
```diff
@@ -56,18 +68,26 @@ pub fn tail<G>(
-                                    results.push((*time, row.clone()));
+                                packer.push(Datum::from(numeric::Numeric::from(*time)));
+                                if connector.emit_progress {
+                                    packer.push(Datum::False);
```
What does this `Datum::False` indicate?

aljoscha replied:
It's about this one: https://materialize.com/docs/sql/tail/#progress

But you're right, I'll add an inline comment explaining this.

JLDLaughlin commented on src/materialized/tests/sql.rs:
```diff
@@ -149,6 +149,37 @@ fn test_time() -> Result<(), Box<dyn Error>> {
+    let mut events = vec![];
+
+    let data = format!("line {}", 42);
+    client_writes.execute(
```
I don't know if this is overkill, but would you want to add a retraction of some sort to this test?

aljoscha replied:
I added a test that outputs negative diffs/multiplicities. 👍

JLDLaughlin commented on src/coord/src/coord.rs:
```diff
@@ -1307,25 +1307,23 @@ impl Coordinator {
-                    if let Some(token) = self.since_handles.get(&id) {
-                        tokens.push(token.clone());
-                    }
+        // For some sinks, we need to block compaction of each timestamp binding
```
This is super helpful!

JLDLaughlin reviewed (approved):
These changes look great to me!

aljoscha commented:
Thanks for the review! I added changes as separate `fixup!...` commits that I will squash before merging. This way it's easier for you to see the change. I also rebased on `main` because this PR was a bit older. There we some conflicts but nothing big.

JLDLaughlin reviewed (approved):
Sorry, I meant to approve the first time! This all looks good to me!

aljoscha commented:
You did approve, but I re-requested a review for the changes. 😅  We always did this for Flink, where we waited for follow-ups to also be approved. But I think Materialize handles this a bit differently.
